lib/libutil/efun.3
.\" $MidnightBSD$
.\"     $NetBSD: efun.3,v 1.4 2006/10/20 17:40:18 christos Exp $
.\"
.\" Copyright (c) 2006 The NetBSD Foundation, Inc.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to The NetBSD Foundation
.\" by Christos Zoulas.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"        This product includes software developed by the NetBSD
.\"        Foundation, Inc. and its contributors.
.\" 4. Neither the name of The NetBSD Foundation nor the names of its
.\"    contributors may be used to endorse or promote products derived
.\"    from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd October 6, 2006
.Dt EFUN 3
.Os
.Sh NAME
.Nm esetfunc ,
.Nm easprintf ,
.Nm efopen ,
.Nm emalloc ,
.Nm ecalloc ,
.Nm erealloc ,
.Nm estrdup ,
.Nm estrlcat ,
.Nm estrlcpy ,
.Nm evasprintf
.Nd error-checked utility functions
.Sh LIBRARY
.Lb libutil
.Sh SYNOPSIS
.In util.h
.Ft void (*)(int, const char *, ...)
.Fn esetfunc "void (*)(int, const char *, ...)"
.Ft int
.Fn easprintf "char ** restrict str" "const char * restrict fmt" "..."
.Ft FILE *
.Fn efopen "const char *p" "const char *m"
.Ft void *
.Fn ecalloc "size_t n" "size_t c"
.Ft void *
.Fn emalloc "size_t n"
.Ft void *
.Fn erealloc "void *p" "size_t n"
.Ft char *
.Fn estrdup "const char *s"
.Ft size_t
.Fn estrlcat "char *dst" "const char *src" "size_t len"
.Ft size_t
.Fn estrlcpy "char *dst" "const char *src" "size_t len"
.Ft int
.Fn evasprintf "char ** restrict str" "const char * restrict fmt" "..."
.Sh DESCRIPTION
The
.Nm easprintf ,
.Nm efopen ,
.Nm ecalloc ,
.Nm emalloc ,
.Nm erealloc ,
.Nm estrdup ,
.Nm estrlcat ,
.Nm estrlcpy ,
and
.Nm evasprintf
functions
operate exactly as the correspoding functions that do not start with an
.Sq e
except that they call
the installed error handler that can be configured with
.Nm esetfunc .
The default error handler is
.Xr err 3 .
The function
.Nm esetfunc
returns the previous error handler function.
A
.Dv NULL
error handler will just call
.Xr exit 3 .
.Sh SEE ALSO
.Xr asprintf 3 ,
.Xr calloc 3 ,
.Xr err 3 ,
.Xr exit 3 ,
.Xr fopen 3 ,
.Xr malloc 3 ,
.Xr realloc 3 ,
.Xr strdup 3 ,
.Xr strlcat 3 ,
.Xr strlcpy 3 ,
.Xr vasprintf 3


NEWS
GNU diffutils NEWS                                    -*- outline -*-

* Noteworthy changes in release 3.2 (2011-09-02) [stable]

** Changes in behavior

  --ignore-file-name-case now applies at the top level too.
  For example, "diff dir inIt" might compare "dir/Init" to "inIt".

** New features

  diff and sdiff have a new option --ignore-trailing-space (-Z).

** Packaging

  The texinfo documentation no longer specifies "front-cover" or "back-cover"
  texts, so that it may now be included in Debian's "main" section.


* Noteworthy changes in release 3.1 (2011-08-10) [stable]

** Bug fixes

  diff no longer reports spurious differences merely because two entries
  in the same directory have names that compare equal in the current
  locale, or compare equal because --ignore-file-name-case was given.

* Noteworthy changes in release 3.0 (2010-05-03) [stable]

** Bug fixes

  diff once again prints the required "\ No newline at end of file" line
  when at least one input lacks a newline-at-EOF and the final hunk plus
  context-length aligns exactly with the end of the newline-lacking file.
  [bug introduced between 2.8.7 and 2.9]

** Changes in behavior

  In context-style diffs, diff prints a portion of a preceding "function"
  line for each hunk, with --show-function-line=RE (-F) or
  --show-c-function (-p).  Now, it trims leading blanks from such lines
  before extracting a prefix.  This is useful especially when a function
  line is so far indented that the name itself would be truncated or not
  included in the limited-width substring that diff appends.

  diff once again reports a difference with the diagnostic
  "Binary files A and B differ" when at least one of the files
  appears to be binary.  From 2.8.4 through diffutils-2.9, it printed
  "Files A and B differ".


* Noteworthy changes in release 2.9 (2010-02-11) [stable]

** New features

  New diff option --suppress-blank-empty.

  Bring back support for `diff -NUM', where NUM is a number,
  even when conforming to POSIX 1003.1-2001.  This change reverts to
  the behavior of GNU diff 2.7 and earlier.  This is a change only
  when conforming to POSIX 1003.1-2001; there is no effect when
  conforming to older POSIX versions.

  This change is in response to decisions taken in the January 2005
  Austin Group standardization meeting.  For more details, please see
  "Utility Syntax Guidelines" in the Minutes of the January 2005
  Meeting <http://www.opengroup.org/austin/docs/austin_239.html>.

  sdiff now understands '1' and '2' as synonyms for 'l' and 'r'.

** Changes in behavior

  sdiff and diff3 now invoke diff, not $(bindir)/diff

** Administrivia

  New discussion and bug-reporting address: bug-diffutils@gnu.org

  updated gnulib support


* Noteworthy changes in release 2.8.7 (2004-04-13) [stable]

  Version 2.8.7 contains no user-visible changes.

User-visible changes in version 2.8.6:

* New diff3 option --strip-trailing-cr.

* With -N and -P, inaccessible empty regular files (the kind of files
  that 'patch' creates to indicate nonexistent backups) are now
  treated as nonexistent when they are in the 'backup' file position.

* If multiple SKIP values are given to cmp, e.g., `cmp -i 10 -i 20',
  cmp now uses the maximal value instead of the last one.

* diff now omits the ".000000000" on hosts that do not support
  fractional time stamps.

Version 2.8.5 was not publicly released.

User-visible changes in version 2.8.4:

* Diff now simply prints "Files A and B differ" instead of "Binary
  files A and B differ".  The message is output if either A or B
  appears to be a binary file, and the old wording was misleading
  because it implied that both files are binary, which is not
  necessarily the case.

User-visible changes in version 2.8.3:

* New locale: en_US.

User-visible changes in version 2.8.2:

* New diff and sdiff option:
  --tabsize=COLUMNS
* If --ignore-space-change or --ignore-all-space is also specified,
  --ignore-blank-lines now considers lines to be empty if they contain
  only white space.
* More platforms now handle multibyte characters correctly when
  excluding files by name (diff -x and -X).
* New locales: hu, pt_BR.

User-visible changes in version 2.8.1:

* Documentation fixes.

User-visible changes in version 2.8:

* cmp and diff now conform to POSIX 1003.1-2001 (IEEE Std 1003.1-2001)
  if the underlying system conforms to POSIX and if the _POSIX2_VERSION
  environment variable is set to 200112.  Conformance removes support
  for `diff -NUM', where NUM is a number.  Use -C NUM or -U NUM instead.
* cmp now supports trailing operands SKIP1 and SKIP2, like BSD cmp.
* cmp -i or --ignore-initial now accepts SKIP1:SKIP2 option value.
* New cmp option: -n or --bytes.
* cmp's old -c or --print-chars option has been renamed;
  use -b or --print-bytes instead.
* cmp now outputs "byte" rather than "char" outside the POSIX locale.
* cmp -l's index column width now adjusts to fit larger (or smaller) files.
* cmp -l -s and cmp -s -l are not allowed.  Use cmp -s or cmp -l instead.
* diff uses ISO 8601 style time stamps for output times (e.g. "2001-11-23
  16:44:36.875702460 -0800") unless in the C or POSIX locale and the
  -c style is specified.
* diff's -I and -F options use the regexp syntax of grep, not of Emacs.
* diff now accepts multiple context arguments, and uses their maximum value.
* New diff and sdiff options:
  -E  --ignore-tab-expansion
  --strip-trailing-cr
* New diff options:
  --from-file=FILE, --to-file=FILE
  --ignore-file-name-case
  --no-ignore-file-name-case
* New diff3 and sdiff option:
  --diff-program=PROGRAM
* The following diff options are still accepted, but are no longer documented.
  They may be withdrawn in future releases.
  -h (omit; it has no effect)
  -H (use --speed-large-files instead)
  -L (use --label instead)
  -P (use --unidirectional-new-file instead)
  --inhibit-hunk-merge (omit; it has no effect)
* Recursive diffs now sort file names according to the LC_COLLATE locale

lib/libutil/efun.c
/* $MidnightBSD$ */
/*	$NetBSD: efun.c,v 1.4 2006/09/27 16:20:03 christos Exp $	*/

/*-
 * Copyright (c) 2006 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is derived from software contributed to The NetBSD Foundation
 * by Christos Zoulas.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *        This product includes software developed by the NetBSD
 *        Foundation, Inc. and its contributors.
 * 4. Neither the name of The NetBSD Foundation nor the names of its
 *    contributors may be used to endorse or promote products derived
 *    from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <sys/cdefs.h>

#include <sys/types.h>
#include <err.h>
#include <errno.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <stdarg.h>
#include <libutil.h>

static void (*efunc)(int, const char *, ...) = err;

void (*
esetfunc(void (*ef)(int, const char *, ...)))(int, const char *, ...)
{
	void (*of)(int, const char *, ...) = efunc;
	efunc = ef == NULL ? (void (*)(int, const char *, ...))exit : ef;
	return of;
}

size_t
estrlcpy(char *dst, const char *src, size_t len)
{
	size_t rv;
	if ((rv = strlcpy(dst, src, len)) >= len) {
		errno = ENAMETOOLONG;
		(*efunc)(1,
		    "Cannot copy string; %zu chars needed %zu provided",
		    rv, len);
	}
	return rv;
}

size_t
estrlcat(char *dst, const char *src, size_t len)
{
	size_t rv;
	if ((rv = strlcat(dst, src, len)) >= len) {
		errno = ENAMETOOLONG;
		(*efunc)(1,
		    "Cannot append to string; %zu chars needed %zu provided",
		    rv, len);
	}
	return rv;
}

char *
estrdup(const char *s)
{
	char *d = strdup(s);
	if (d == NULL)
		(*efunc)(1, "Cannot copy string");
	return d;
}

void *
emalloc(size_t n)
{
	void *p = malloc(n);
	if (p == NULL)
		(*efunc)(1, "Cannot allocate %zu bytes", n);
	return p;
}

void *
ecalloc(size_t n, size_t s)
{
	void *p = calloc(n, s);
	if (p == NULL)
		(*efunc)(1, "Cannot allocate %zu bytes", n);
	return p;
}

void *
erealloc(void *p, size_t n)
{
	void *q = realloc(p, n);
	if (q == NULL)
		(*efunc)(1, "Cannot re-allocate %zu bytes", n);
	return q;
}

FILE *
efopen(const char *p, const char *m)
{
	FILE *fp = fopen(p, m);
	if (fp == NULL)
		(*efunc)(1, "Cannot open `%s'", p);
	return fp;
}

#ifdef __BSD_VISIBLE

int
easprintf(char **restrict ret, const char *restrict format, ...)
{
	int rv;
	va_list ap;
	va_start(ap, format);
	if ((rv = vasprintf(ret, format, ap)) == -1)
		(*efunc)(1, "Cannot format string");
	va_end(ap);
	return rv;
}

int
evasprintf(char **restrict ret, const char *restrict format, va_list ap)
{
	int rv;
	if ((rv = vasprintf(ret, format, ap)) == -1)
		(*efunc)(1, "Cannot format string");
	return rv;
}

#endif
